addons/website_slides/models/slide_channel.py

```diff
@@ -104,24 +105,21 @@ def _set_as_completed(self, completed=True):
             True if we make the slide as completed
             False if we remove user completion
         """
-        partner_karma = dict.fromkeys(self.mapped('partner_id').ids, 0)
-        for record in self:
-            record.completed = completed
-            partner_karma[record.partner_id.id] += record.channel_id.karma_gen_channel_finish
-
-        partner_karma = {
-            partner_id: karma_to_add
-            for partner_id, karma_to_add in partner_karma.items() if karma_to_add > 0
-        }
-
-        if partner_karma:
-            users = self.env['res.users'].sudo().search([('partner_id', 'in', list(partner_karma.keys()))])
-            for user in users:
-                karma = partner_karma[user.partner_id.id]
-                if not completed:
-                    # Mark the channel as not-completed, we remove the gained karma
-                    karma *= -1
-                users.add_karma(karma)
+        for channel, partners in groupby(self, lambda partner: partner.channel_id):
+            partners = reduce(lambda a, b: a | b, partners)
+            partners.completed = completed
+            karma = channel.karma_gen_channel_finish
+            if karma > 0:
+                users = self.env['res.users'].sudo().search([('partner_id', 'in', partners.partner_id.ids)])
```

Comment:
Does one search per loop, but I guess this is acceptable to simplify (and it allows testing if we add karma).